resolver/addressable_resolver.go

```diff
@@ -56,6 +58,28 @@ func NewURIResolver(ctx context.Context, callback func(types.NamespacedName)) *U
 		},
 	}
 
+	ret.listerFactory = func(gvr schema.GroupVersionResource) (cache.GenericLister, error) {
+		_, l, err := ret.informerFactory.Get(ctx, gvr)
+		return l, err
+	}
+
+	return ret
+}
+
+// NewURIResolver constructs a new URIResolver with context and a callback
+// for a given listableType (Listable) passed to the URIResolver's tracker.
+// extra when controller context is used
```

Comment:
I'd like to see this fleshed out into a full sentence.. WDYT of something like:

~~~~
// This method returns a URIResolver which uses the passed context to retrieve
// listerFactories, and should be used when the URIResolver will be reused for multiple reconciles.
~~~~

.. although writing this out makes me wonder if there's any use case where you _shouldnt_ use the new version, and if so whether there's a good reason not to update the behaviour of the original `NewURIResolver` method instead 🤔 

ping @n3wscott for thoughts on that cos I dont have that much context on URIResolver use outside of domainmapping (which is where we use it in serving).

Reply:
Yeah, the uri resolver is used in trigger, pingsource, subscribtion and apiserversource reconcilers in eventing. Also discussed that in serving-api wg and we thought we'll have the context timeout in genreconciler as an opt in. That's why I created a new constructor.

Reply:
I guess Im wondering if there'd be a downside to having NewURIResolver _always_ set things up to use the constructor context in the resolve calls when looking up the factory. That wouldn't be needed if you don't set a timeout on the context you pass in the reconcile, but it wouldn't (I think?) hurt in that case, and it seems maybe more correct anyway not to leak the per-reconcile context outside of the reconcile (plus avoids having two different code paths/constructors iyswim). 

It's hard to know when writing a reconciler if someone will later decide to add a timeout, so seems nice if things would work "just work" if they do without needing to remember to use the right constructor.

To be clear Im not saying there's definitely not a reason to keep the existing behaviour and add a new constructor for the new behaviour... I just can't think of one and if there _isn't_ it might be nice to be able to avoid adding surface area to the api. (paging @n3wscott to the courtesy telephone again 😊)